raft/src/main/java/org/apache/kafka/raft/RaftUtil.java

```diff
@@ -95,11 +107,417 @@ public static FetchResponseData singletonFetchResponse(
+                    .setPartitions(Collections.singletonList(
+                        new VoteResponseData.PartitionData()
+                            .setErrorCode(partitionLevelError.code())
+                            .setLeaderId(leaderId)
+                            .setLeaderEpoch(leaderEpoch)
+                            .setVoteGranted(voteGranted)))));
+
+        if (apiVersion >= 1) {
+            Optional<InetSocketAddress> address = endpoints.address(listenerName);
+            if (address.isPresent() && leaderId >= 0) {
+                // Populate the node endpoints
+                VoteResponseData.NodeEndpointCollection nodeEndpoints = new VoteResponseData.NodeEndpointCollection(1);
+                nodeEndpoints.add(
+                    new VoteResponseData.NodeEndpoint()
+                        .setNodeId(leaderId)
+                        .setHost(address.get().getHostString())
+                        .setPort(address.get().getPort())
+                );
+                response.setNodeEndpoints(nodeEndpoints);
+            }
+        }
+
+        return response;
+    }
+
+    public static FetchSnapshotRequestData singletonFetchSnapshotRequest(
+        String clusterId,
+        ReplicaKey replicaKey,
+        TopicPartition topicPartition,
+        int epoch,
+        OffsetAndEpoch offsetAndEpoch,
+        int maxBytes,
+        long position
+    ) {
+        FetchSnapshotRequestData.SnapshotId snapshotId = new FetchSnapshotRequestData.SnapshotId()
+            .setEndOffset(offsetAndEpoch.offset())
+            .setEpoch(offsetAndEpoch.epoch());
+
+        FetchSnapshotRequestData.PartitionSnapshot partitionSnapshot = new FetchSnapshotRequestData.PartitionSnapshot()
+            .setPartition(topicPartition.partition())
+            .setCurrentLeaderEpoch(epoch)
+            .setSnapshotId(snapshotId)
+            .setPosition(position)
+            .setReplicaDirectoryId(replicaKey.directoryId().orElse(ReplicaKey.NO_DIRECTORY_ID));
+
+        return new FetchSnapshotRequestData()
+            .setClusterId(clusterId)
+            .setReplicaId(replicaKey.id())
+            .setMaxBytes(maxBytes)
+            .setTopics(
+                Collections.singletonList(
+                    new FetchSnapshotRequestData.TopicSnapshot()
+                        .setName(topicPartition.topic())
+                        .setPartitions(Collections.singletonList(partitionSnapshot))
+                )
+            );
+    }
+
+    /**
+     * Creates a FetchSnapshotResponseData with a single PartitionSnapshot for the topic partition.
+     *
+     * The partition index will already be populated when calling operator.
+     *
+     * @param topicPartition the topic partition to include
+     * @param operator unary operator responsible for populating all of the appropriate fields
+     * @return the created fetch snapshot response data
+     */
+    public static FetchSnapshotResponseData singletonFetchSnapshotResponse(
```

Comment:
Fixed the Java Doc.

This method has a java doc because I copied it from `FetchSnapshotResponse` when I moved the method. Java doc are nice to help others understand the code. Maybe the person working on https://issues.apache.org/jira/browse/KAFKA-16928 can add the Java Docs.